packages/kbn-unified-data-table/src/components/data_table.tsx

```diff
@@ -603,74 +604,42 @@ export const UnifiedDataTable = ({
     typeof rowsPerPageState === 'number' && rowsPerPageState > 0
       ? rowsPerPageState
       : DEFAULT_ROWS_PER_PAGE;
-  const [pagination, setPagination] = useState({
-    pageIndex: 0,
-    pageSize: currentPageSize,
-  });
+  const [currentPageIndex, setCurrentPageIndex] = useState(0);
   const rowCount = useMemo(() => (displayedRows ? displayedRows.length : 0), [displayedRows]);
   const pageCount = useMemo(
-    () => Math.ceil(rowCount / pagination.pageSize),
-    [rowCount, pagination]
+    () => Math.ceil(rowCount / currentPageSize),
+    [currentPageSize, rowCount]
   );
 
-  const paginationObj = useMemo(() => {
+  const pagination = useMemo<EuiDataGridPaginationProps | undefined>(() => {
     const onChangeItemsPerPage = (pageSize: number) => {
       onUpdateRowsPerPage?.(pageSize);
     };
 
     const onChangePage = (pageIndex: number) => {
-      setPagination((paginationData) => ({ ...paginationData, pageIndex }));
+      setCurrentPageIndex(pageIndex);
+      onUpdatePageIndex?.(pageIndex);
     };
 
     return isPaginationEnabled
       ? {
           onChangeItemsPerPage,
           onChangePage,
-          pageIndex: pagination.pageIndex > pageCount - 1 ? 0 : pagination.pageIndex,
-          pageSize: pagination.pageSize,
-          pageSizeOptions: rowsPerPageOptions ?? getRowsPerPageOptions(pagination.pageSize),
+          pageIndex: currentPageIndex > pageCount - 1 ? 0 : currentPageIndex,
```

Comment:
```suggestion
          pageIndex: currentPageIndex,
```
...Then passing `currentPageIndex` directly here.

And we'd only need to manually sync when `pageCount` changes in a separate effect:
```ts
useEffect(() => {
  setCurrentPageIndex((previousPageIndex) => {
    const calculatedPageIndex = previousPageIndex > pageCount - 1 ? 0 : previousPageIndex;

    if (calculatedPageIndex !== previousPageIndex) {
      onUpdatePageIndex?.(calculatedPageIndex);
    }

    return calculatedPageIndex;
  });
}, [onUpdatePageIndex, pageCount]);
```